server/protocols/webadmin-integration-test/src/test/java/org/apache/james/webadmin/integration/EventDeadLettersIntegrationTest.java

```diff
@@ -0,0 +1,464 @@
+        .post(EventDeadLettersRoutes.BASE_PATH + "/groups/" + SERIALIZED_GROUP)
+            .jsonPath()
+            .get("taskId");
+
+        given()
+            .basePath(TasksRoutes.BASE)
+        .when()
+            .get(taskId + "/await")
+        .then()
+            .body("status", is("completed"))
+            .body("additionalInformation.successfulRedeliveriesCount", is(2))
+            .body("additionalInformation.failedRedeliveriesCount", is(0))
+            .body("additionalInformation.group", is(SERIALIZED_GROUP));
+
+        assertThat(retryEventsListener.getSuccessfulEvents()).containsExactlyInAnyOrder(failedEvent1, failedEvent2);
+    }
+
+    @Test
+    public void multipleFailedEventsShouldNotBeInDeadLettersAfterSuccessfulAllRedelivery() {
+        retryEventsListener.setRetriesBeforeSuccess(4);
+        mailboxProbe.createMailbox("#private", BOB, DefaultMailboxes.INBOX);
+        mailboxProbe.createMailbox("#bobox", BOB, DefaultMailboxes.INBOX);
+
+        String taskId = with()
+            .queryParam("action", EVENTS_ACTION)
+        .post(EventDeadLettersRoutes.BASE_PATH)
+            .jsonPath()
+            .get("taskId");
+
+        given()
+            .basePath(TasksRoutes.BASE)
+        .when()
+            .get(taskId + "/await")
+        .then()
+            .body("status", is("completed"))
+            .body("additionalInformation.successfulRedeliveriesCount", is(2))
+            .body("additionalInformation.failedRedeliveriesCount", is(0));
+
+        when()
+            .get(EventDeadLettersRoutes.BASE_PATH + "/groups/" + SERIALIZED_GROUP)
+        .then()
+            .statusCode(HttpStatus.OK_200)
+            .contentType(ContentType.JSON)
+            .body(".", hasSize(0));
+    }
+
+    @Test
+    public void multipleFailedEventsShouldBeCorrectlyProcessedByListenerAfterSuccessfulAllRedelivery() {
+        retryEventsListener.setRetriesBeforeSuccess(4);
+        mailboxProbe.createMailbox("#private", BOB, DefaultMailboxes.INBOX);
+        mailboxProbe.createMailbox("#bobox", BOB, DefaultMailboxes.INBOX);
+
+        Event failedEvent1 = retryEventsListener.getFailedEvents().get(0);
+        Event failedEvent2 = retryEventsListener.getFailedEvents().get(1);
+
+        String taskId = with()
+            .queryParam("action", EVENTS_ACTION)
+        .post(EventDeadLettersRoutes.BASE_PATH)
+            .jsonPath()
+            .get("taskId");
+
+        given()
+            .basePath(TasksRoutes.BASE)
+        .when()
+            .get(taskId + "/await")
+        .then()
+            .body("status", is("completed"))
+            .body("additionalInformation.successfulRedeliveriesCount", is(2))
+            .body("additionalInformation.failedRedeliveriesCount", is(0));
+
+        assertThat(retryEventsListener.getSuccessfulEvents()).containsExactlyInAnyOrder(failedEvent1, failedEvent2);
+    }
+
+    @Test
+    public void failedEventShouldStillBeInDeadLettersAfterFailedRedelivery() {
```

Comment:
Fails for now. Seems to be that cannot bring back up as an error if redeliver exceeds max number of retries. Need to find the issue with the task (might be the fact that I do the remove after a "success"... maybe should remove first, if redeliver doesn't work after then it will put the event back to the event dead letters)

Reply:
We need to rework dead-letter to account that an event can be inserted several time.

The caller should explicitly specify which event he wishes to delete.

This arises with the InVmEventBus as synchronous processing is allowed.

I will open a ticket regarding this.